### ISSUE TITLE
Relax migration ordering semantics

### ISSUE THREAD

**Issue by k0001** [k0001/umzug#3]
The rationale for this is that collaboration in large teams gets tricky if more than one person is adding migrations at the same time, seeing how they need to be ordered before merging.

We can remove the restriction of migration order if we mandate that each migration lists zero or more other migrations as its dependencies. Then, if migration `B` depends on migration `A`, and migration `C` also depends on migration `A`, then any permutation of the migration list `[A, B, C]` will work, with `A` being run first, and both `B` and `C` being run afterwards (no particular preference as to which of those two runs first).

**Comment by imalsogreg** [comment]
There are several places where the ordering of migrations is implicit in the List index of the migrations. If Migrations themselves list their dependencies, should migrations be collected in a `Set` instead, to avoid the confusion that the List sets the order?

**Comment by k0001** [comment]
Yes, good idea. We could fallback to the list ordering, but this would make it tempting to just not  being explicit about dependencies, which will have a cost down the road.
